README.md
# housing-prices ![](https://circleci.com/gh/dev-daeun/housing-prices.svg?style=svg)

Flask application in this repo provides housing prices prediction in Boston.

Model data & flask application is provided by [udacity/DevOps_Microservices](https://github.com/udacity/DevOps_Microservices)

## API specification
```
* Request *
POST localhost:8000/predict/

{  
   "CHAS":{  
      "0":0
   },
   "RM":{  
      "0":6.575
   },
   "TAX":{  
      "0":296.0
   },
   "PTRATIO":{  
      "0":15.3
   },
   "B":{  
      "0":396.9
   },
   "LSTAT":{  
      "0":4.98
   }
}


* Response *
{
  "prediction": [
    20.35373177134412
  ]
}

```


## How to setup
If you want to run Flask application server in your host, execute commands below.
```
$ cd ~/housing-prices
$ make setup
$ make install
$ python app.py
$ sh make_prediction.sh
```
If you want to run Flask application server in docker container, execute commands below.
```
$ sh run_docker.sh
$ sh make_prediction.sh
```

If you want to run Flask application server in Kubernetes cluster, execute commands below.
(Minikube should be installed and started on your local PC.)
```
# Run the Docker Hub container with kubernetes
$ kubectl create deployment housing-prices --image=kde6260/housing-prices:latest
```
```
# List kubernetes pods
$ kubectl get pods | grep housing-prices
$ export POD_NAME=$(kubectl get pods -o go-template --template '{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}')
```
```
# Forward the container port to a host
kubectl port-forward $POD_NAME 8000:80
```
```
$ Call API
$ sh make_prediction.sh
```


CHANGELOG.md
# CHANGELOG



## v0.1.0 (2023-11-02)

### Unknown

* test for tag ([`b4e4630`](https://github.com/dev-daeun/housing-prices/commit/b4e46308de5cc586b92e7bb40a4948b10356d8ba))

* test ([`274c448`](https://github.com/dev-daeun/housing-prices/commit/274c448bf1fd0187c7e280dec7a6a98dfad6c4e1))


## v0.0.47 (2020-01-27)

### Unknown

* log status of pod and response from application ([`1cd3c4b`](https://github.com/dev-daeun/housing-prices/commit/1cd3c4b2ce3decf3b4ce0dc0adb4d3c2982f73ea))

* identify source of model data and application code ([`e316dc7`](https://github.com/dev-daeun/housing-prices/commit/e316dc78cb4aafbd3e2acb509583f021fce86fbd))

* move output files into directory ([`6365fab`](https://github.com/dev-daeun/housing-prices/commit/6365fab71f87bc30bcc5d3233b1dd519a58c4246))

* fix tag of image in dockerhub ([`ae3bdf7`](https://github.com/dev-daeun/housing-prices/commit/ae3bdf79ce0bf0174070f8c4ca1cde884d421d94))

* log final result of calculating prediction ([`e7dfbd9`](https://github.com/dev-daeun/housing-prices/commit/e7dfbd90c7a159c401a18a2c90ead7dbbb07a567))

* update README.md ([`7c36d09`](https://github.com/dev-daeun/housing-prices/commit/7c36d0905045685d95fefb6fd08a26ee1f40d255))

* add circleci script ([`7452327`](https://github.com/dev-daeun/housing-prices/commit/7452327ab1561d8ea8c7c948555bc96dd9734e43))

* use % format in log string ([`cad3027`](https://github.com/dev-daeun/housing-prices/commit/cad3027aaccc8eca26cf17b83d6d1dabdd0a098d))

* remove unused instruction &amp; install wheel ([`46168ab`](https://github.com/dev-daeun/housing-prices/commit/46168ab36bd6b2e9d2be7332761aebb5abc3543c))

* change environment name ([`afd4ee2`](https://github.com/dev-daeun/housing-prices/commit/afd4ee28b9657aa1b9c28108fd39245e49cc1991))

* rewrite run_kubernetes.sh to create pod &amp; foward pod to the host ([`02c5878`](https://github.com/dev-daeun/housing-prices/commit/02c5878757ba3a72d9c2c8e4d6b3b5825f0883cb))

* tag &amp; push an image to dockerhub repo ([`7d43702`](https://github.com/dev-daeun/housing-prices/commit/7d4370231f087b63ac806a2973cafb5decd5b981))

* create logfile ([`5588211`](https://github.com/dev-daeun/housing-prices/commit/5588211fe34eb7594fe910d52b714dae85152b09))

* write explicit version of each package ([`f3fc6d5`](https://github.com/dev-daeun/housing-prices/commit/f3fc6d5f17c70ac8ae494704c45260714d835c08))

* rewrite run_docker.sh to build images &amp; run container ([`9b87834`](https://github.com/dev-daeun/housing-prices/commit/9b878349cd36a0f05cbe7349dc717e1bd53adc4a))

* remove unnecessary &amp; not usable sudo option ([`281e831`](https://github.com/dev-daeun/housing-prices/commit/281e83106871415916eeea711822702085d7f488))

* fix Makefile to install python packages from pypi.org only ([`5d19f0e`](https://github.com/dev-daeun/housing-prices/commit/5d19f0e7bb09505ef262dbd0503d6963b254c274))

* rewrite Dockerfile to run flask server ([`8841c49`](https://github.com/dev-daeun/housing-prices/commit/8841c49eedb6331fa5e8d2c055cfbfb2b9baab86))

* lint app.py &amp; Makefile ([`e8296b8`](https://github.com/dev-daeun/housing-prices/commit/e8296b8b10ccc37ce8e35eeefbd82a963c492f74))

* create .gitignore ([`a7ed6e1`](https://github.com/dev-daeun/housing-prices/commit/a7ed6e197b196c5bb2885c92ece244c2b701f3e9))

* add fundamental files ([`873d41e`](https://github.com/dev-daeun/housing-prices/commit/873d41ec70d456605671b372b4e7761e11ca9979))

* Initial commit ([`645afae`](https://github.com/dev-daeun/housing-prices/commit/645afae68756e818f49ef9697c6ad32bcce18610))


app.py
from flask import Flask, request, jsonify
from flask.logging import create_logger
import logging

import pandas as pd
from sklearn.externals import joblib
from sklearn.preprocessing import StandardScaler

app = Flask(__name__)
LOG = create_logger(app)
LOG.setLevel(logging.INFO)


def scale(payload):
    """Scales Payload"""

    LOG.info('Scaling Payload: \n%s', payload)
    scaler = StandardScaler().fit(payload.astype(float))
    scaled_adhoc_predict = scaler.transform(payload.astype(float))
    return scaled_adhoc_predict


@app.route("/")
def home():
    html = f"<h3>Sklearn Prediction Home</h3>"
    return html.format(format)


@app.route("/predict", methods=['POST'])
def predict():
    """Performs an sklearn prediction
        
        input looks like:
        {
        "CHAS":{
        "0":0
        },
        "RM":{
        "0":6.575
        },
        "TAX":{
        "0":296.0
        },
        "PTRATIO":{
        "0":15.3
        },
        "B":{
        "0":396.9
        },
        "LSTAT":{
        "0":4.98
        }
        
        result looks like:
        { "prediction": [ <val> ] }
        
        """

    # Logging the input payload
    json_payload = request.json
    LOG.info('JSON payload: \n%s', json_payload)
    inference_payload = pd.DataFrame(json_payload)
    LOG.info('Inference payload DataFrame: \n%s', inference_payload)
    # scale the input
    scaled_payload = scale(inference_payload)
    # get an output prediction from the pretrained model, clf
    prediction = list(clf.predict(scaled_payload))
    LOG.info('output preduction: %s', prediction)
    # TO DO:  Log the output prediction value
    return jsonify({'prediction': prediction})


if __name__ == "__main__":
    # load pretrained model as clf
    clf = joblib.load("./model_data/boston_housing_prediction.joblib")
    app.run(host='0.0.0.0', port=80, debug=True)  # specify port=80
